homeassistant/components/notify/__init__.py

```diff
@@ -4,110 +4,281 @@
-    vol.Optional(ATTR_TITLE, default=ATTR_TITLE_DEFAULT): cv.string,
+    vol.Optional(ATTR_TITLE, default=ATTR_TITLE_DEFAULT): cv.template,
     vol.Optional(ATTR_TARGET): cv.string,
-    vol.Optional(ATTR_DATA): dict,
+    vol.Optional(ATTR_FORMATS, default={}): dict,
 })
 
+PHOTO_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+    vol.Optional(ATTR_URL): cv.string,
+    vol.Optional(ATTR_FILE): cv.string,
+    vol.Optional(ATTR_CAMERA): cv.template,
+    vol.Optional(ATTR_CAPTION): cv.string,
+    vol.Optional(ATTR_USERNAME): cv.string,
+    vol.Optional(ATTR_PASSWORD): cv.string,
+})
+
+LOCATION_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+    vol.Optional(ATTR_LATITUDE): cv.latitude,
+    vol.Optional(ATTR_LONGITUDE): cv.longitude,
+    vol.Optional(ATTR_DEVICE): cv.string,
+    vol.Optional(ATTR_CAPTION): cv.template,
+})
+
+
 _LOGGER = logging.getLogger(__name__)
 
 
-def send_message(hass, message, title=None):
+# pylint: disable=too-many-arguments
+def send_message(hass, message=None, entity_id=None, title=None, target=None,
+                 formats=None):
     """Send a notification message."""
     data = {
-        ATTR_MESSAGE: message
+        key: value for key, value in [
+            (ATTR_ENTITY_ID, entity_id),
+            (ATTR_TITLE, title),
+            (ATTR_MESSAGE, message),
+            (ATTR_TARGET, target),
+            (ATTR_FORMATS, formats)
+        ] if value is not None
     }
 
-    if title is not None:
-        data[ATTR_TITLE] = title
+    hass.services.call(DOMAIN, SERVICE_SEND_MESSAGE, data)
 
-    hass.services.call(DOMAIN, SERVICE_NOTIFY, data)
+
+# pylint: disable=too-many-arguments
+def send_photo(hass, entity_id=None, url=None, file=None, camera_entity=None,
+               caption=None, username=None, password=None):
+    """Send a photo."""
+    data = {
+        key: value for key, value in [
+            (ATTR_ENTITY_ID, entity_id),
+            (ATTR_URL, url),
+            (ATTR_FILE, file),
+            (ATTR_CAMERA, camera_entity),
+            (ATTR_CAPTION, caption),
+            (ATTR_USERNAME, username),
+            (ATTR_PASSWORD, password)
+        ] if value is not None
+    }
+
+    hass.services.call(DOMAIN, SERVICE_SEND_PHOTO, data)
+
+
+# pylint: disable=too-many-arguments
+def send_location(hass, entity_id=None, latitude=None, longitude=None,
+                  device_entity=None, caption=None):
+    """Send a location."""
+    data = {
+        key: value for key, value in [
+            (ATTR_ENTITY_ID, entity_id),
+            (ATTR_LATITUDE, latitude),
+            (ATTR_LONGITUDE, longitude),
+            (ATTR_DEVICE, device_entity),
+            (ATTR_CAPTION, caption)
+        ] if value is not None
+    }
+
+    hass.services.call(DOMAIN, SERVICE_SEND_LOCATION, data)
 
 
 def setup(hass, config):
     """Setup the notify services."""
-    success = False
+    component = EntityComponent(
```

Comment:
I think that we need a new helper: `ServiceComponent`. We should not create entities just because we can re-use a piece of code. 

I liked the approach that the existing notify component took: register service by service name. If we can be smart about which methods are on the returned notify `service` object, we can register only the HA services that are supported.

```python
# pseudo code
notify_service = notify_implementation.get_service(hass, p_config)

if hasattr(notify_service, 'send_photo'):
    register_service(service_base_photo, notify_service)

if hasattr(notify_service, 'send_message'):
    register_service(service_base_message, notify_service)
```